### ISSUE TITLE
How to calculate the max HeapMaxSize given the physical memory size?

### ISSUE THREAD

**Issue by huxh10** [01org/linux-sgx#87]
I have a server with 64GB memory. When I set HeapMaxSize to 0x800000000 (2 ** 35 = 32GB), the compilation succeeds while running the program fails with unloadable enclave.

When HeapMaxSize is 0x700000000 (2 ** 32 * 7 = 28 GB), the program works. Also, I notice the program consumes around 40GB cache memory using top cmd. The TCSNum is 1.

So my question is what is the max value of HeapMaxSize?

**Comment by andyzyb** [comment]
The max size of Enclave is 64GB per spec which includes heap, stack, code/text segments, etc.   The failure of loading may due to different reasons, like lack of continuous linear address of more than 32G, etc.    You probably can debug into uRTS code to see where it fails.

The max EPC size on the platform may be just ~100MB,  so the driver will encrypt and page out the EPC to system memory when it exceeds the EPC size.

**Comment by yuyuany** [comment]
Yes, there's a problem on loading 64GB enclave with current version. It will be fixed in next release.

**Comment by huxh10** [comment]
Thank you!

**Comment by yuyuany** [comment]
I found we've already have pull requests merged for loading large enclave. Can you try the latest code to see if the problem has been fixed?

**Comment by huxh10** [comment]
I pulled the newest version. I set HeapMaxSize 0xB00000000 (2 ** 32 * 11 = 44GB). It worked. I then set HeapMaxSize 0xF00000000 (2 ** 32 * 15 = 60 GB). Enclave creation doesn't fail but doesn't finish either. It has run more than 12 hours.

**Comment by yuyuany** [comment]
Yes, I can reproduce the problem. It looks to be the bug in SGX driver. We will continue to investigate it.

**Comment by andyzyb** [comment]
64GB may be difficult to support, as driver still needs some EPC to manage Enclave paging. On a system with limited EPC (tens of MBs), it may result in pages swap in/out constantly, and not successfully initialize it.

**Comment by jsakkine-intel** [comment]
64GB needs 128 MB of VA pages i.e. the EPC size. Because the driver does not support ATM VA page swapping the allocator ends up waiting forever for a new page.

Ultimate solution is to implement VA swapping but short term fix would be to keep track of number of VA  pages allocated and and fail in sgx_alloc_page with ENOMEM in the case when VA pages take all of EPC. That would at least render the deadlock away.

**Comment by huxh10** [comment]
I have read references about limited EPC size. I am curious why EPC size is fixed? Is it possible to config EPC size through BIOS.

**Comment by rodolfoams** [comment]
@huxh10 I have a Dell Optiplex 5040 and its BIOS allows me to configure the EPC size. I can choose between 32, 64 and 128MB. The EPC size is limited because it is completely reserved for enclave pages, therefore you cannot use any of the EPC for other purposes.

**Comment by huxh10** [comment]
Is there an upper bound for EPC size? How large can EPC size be configured?

**Comment by rodolfoams** [comment]
As of today, the upper bound is 128MB.

**Comment by huxh10** [comment]
This is what I want to ask, is the upper bound fixed? I know that other application can not use EPC, but what if the secure applications want to use more EPC, page swapping decreases the performance a lot.

**Comment by andyzyb** [comment]
Yes, 128 MB is the current max size for EPC.

**Comment by 2channelkrt** [comment]
> 
> 
> > Yes, 128 MB is the current max size for EPC.
> 
> Is there a technical reason for which this is the current max size of the EPC ?

This might not be exact answer, but here's additional info that I know.

SGX encrypts the memory, and tracks its version using hash tree. For hash tree, only lowest level (root) is kept in the dedicated SRAM which is located in the die. From CPU perspective, this is the only place where data is trusted, and all other levels in the hash tree is stored in the 'untrusted' DRAM area. I guess they are limiting the EPC size to 128MB because of the size of SRAM they can squeeze into the die.

You could say that increasing the level of the hash tree can give you more EPC capacity. Well, who knows. Maybe overhead becomes too big.

**Comment by rodolfoams** [comment]
Just adding more info, there are already CPUs with more than 128MB of EPC available.

Azure's "Standard_DC8_v2" VM flavor has 168MB, which would indicate that the server where it is running might have even more EPC available.

**Comment by DylanWangWQF** [comment]
So if I run a program inside the enclave, a lot of class objects are created at runtime, in this case, how do we calculate the used memory? The class objects would be stored in heap memory? (Exceed 128MB)
My enclave image is around 50MB.
